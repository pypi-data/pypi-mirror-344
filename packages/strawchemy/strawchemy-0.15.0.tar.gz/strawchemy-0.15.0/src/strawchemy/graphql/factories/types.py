from __future__ import annotations

import dataclasses
from collections.abc import Generator
from typing import TYPE_CHECKING, Any, TypeVar, override

from strawchemy.dto.backend.dataclass import DataclassDTOBackend
from strawchemy.dto.base import (
    DTOBackend,
    DTOBase,
    DTOFactory,
    DTOFieldDefinition,
    ModelFieldT,
    ModelT,
    Relation,
)
from strawchemy.dto.types import DTO_MISSING, DTOConfig, Purpose
from strawchemy.graph import Node
from strawchemy.graphql import typing as strawchemy_typing
from strawchemy.graphql.constants import AGGREGATIONS_KEY, NODES_KEY
from strawchemy.graphql.dto import (
    AggregateDTO,
    AggregateFieldDefinition,
    DTOKey,
    EnumDTO,
    FunctionFieldDefinition,
    GraphQLFieldDefinition,
)
from strawchemy.graphql.typing import AggregateDTOT, GraphQLDTOT

from .aggregations import AggregationInspector
from .enum import EnumDTOFactory

if TYPE_CHECKING:
    from collections.abc import Generator

    from strawchemy.graph import Node
    from strawchemy.graphql.inspector import GraphQLInspectorProtocol


__all__ = ("AggregateDTOFactory", "DistinctOnFieldsDTOFactory", "RootAggregateTypeDTOFactory", "TypeDTOFactory")

T = TypeVar("T")

_TYPING_NS = vars(strawchemy_typing)


class GraphQLDTOFactory(DTOFactory[ModelT, ModelFieldT, GraphQLDTOT]):
    inspector: GraphQLInspectorProtocol[Any, ModelFieldT]

    def type_description(self) -> str:
        return "GraphQL type"

    @override
    def type_hint_namespace(self) -> dict[str, Any]:
        return super().type_hint_namespace() | _TYPING_NS

    @override
    def iter_field_definitions(
        self,
        name: str,
        model: type[T],
        dto_config: DTOConfig,
        base: type[DTOBase[ModelT]] | None,
        node: Node[Relation[ModelT, GraphQLDTOT], None],
        raise_if_no_fields: bool = False,
        *,
        field_map: dict[DTOKey, GraphQLFieldDefinition[Any, Any]] | None = None,
        **kwargs: Any,
    ) -> Generator[DTOFieldDefinition[ModelT, ModelFieldT], None, None]:
        field_map = field_map if field_map is not None else {}
        for field in super().iter_field_definitions(name, model, dto_config, base, node, raise_if_no_fields, **kwargs):
            key = DTOKey.from_dto_node(node)
            graphql_field = GraphQLFieldDefinition.from_field(field)
            yield graphql_field
            field_map[key + field.name] = graphql_field

    @override
    def factory(
        self,
        model: type[T],
        dto_config: DTOConfig,
        base: type[Any] | None = None,
        name: str | None = None,
        parent_field_def: DTOFieldDefinition[ModelT, ModelFieldT] | None = None,
        current_node: Node[Relation[Any, GraphQLDTOT], None] | None = None,
        raise_if_no_fields: bool = False,
        backend_kwargs: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> type[GraphQLDTOT]:
        field_map: dict[DTOKey, GraphQLFieldDefinition[Any, Any]] = {}
        dto = super().factory(
            model,
            dto_config,
            base,
            name,
            parent_field_def,
            current_node,
            raise_if_no_fields,
            backend_kwargs,
            field_map=field_map,
            **kwargs,
        )
        if not dto.__strawchemy_field_map__:
            dto.__strawchemy_field_map__ = field_map
        dto.__strawchemy_description__ = self.type_description()
        return dto


class TypeDTOFactory(GraphQLDTOFactory[ModelT, ModelFieldT, GraphQLDTOT]):
    def __init__(
        self,
        inspector: GraphQLInspectorProtocol[Any, ModelFieldT],
        backend: DTOBackend[GraphQLDTOT],
        handle_cycles: bool = True,
        type_map: dict[Any, Any] | None = None,
        aggregation_factory: AggregateDTOFactory[ModelT, ModelFieldT, AggregateDTOT] | None = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(inspector, backend, handle_cycles, type_map, **kwargs)
        self._aggregation_factory = aggregation_factory or AggregateDTOFactory(
            inspector, DataclassDTOBackend(AggregateDTO)
        )

    def _aggregation_field(
        self, field_def: DTOFieldDefinition[ModelT, ModelFieldT], dto_config: DTOConfig
    ) -> GraphQLFieldDefinition[ModelT, ModelFieldT]:
        related_model = self.inspector.relation_model(field_def.model_field)
        aggregate_dto_config = dataclasses.replace(dto_config, annotation_overrides={})
        dto = self._aggregation_factory.factory(
            model=related_model, dto_config=aggregate_dto_config, parent_field_def=field_def
        )
        return AggregateFieldDefinition(
            dto_config=dto_config,
            model=dto.__dto_model__,  # pyright: ignore[reportGeneralTypeIssues]
            _model_field=field_def.model_field,
            model_field_name=f"{field_def.name}_aggregate",
            type_hint=dto,
            related_dto=dto,
        )

    @override
    def dto_name(
        self, base_name: str, dto_config: DTOConfig, node: Node[Relation[Any, GraphQLDTOT], None] | None = None
    ) -> str:
        return f"{base_name}{'Input' if dto_config.purpose is Purpose.WRITE else ''}Type"

    @override
    def iter_field_definitions(
        self,
        name: str,
        model: type[T],
        dto_config: DTOConfig,
        base: type[DTOBase[ModelT]] | None,
        node: Node[Relation[ModelT, GraphQLDTOT], None],
        raise_if_no_fields: bool = False,
        *,
        aggregations: bool = False,
        field_map: dict[DTOKey, GraphQLFieldDefinition[Any, Any]] | None = None,
        **kwargs: Any,
    ) -> Generator[DTOFieldDefinition[ModelT, ModelFieldT], None, None]:
        field_map = field_map if field_map is not None else {}
        for field in super().iter_field_definitions(
            name, model, dto_config, base, node, raise_if_no_fields, field_map=field_map, **kwargs
        ):
            key = DTOKey.from_dto_node(node)
            if field.is_relation and field.uselist and aggregations:
                aggregation_field = self._aggregation_field(field, dto_config)
                field_map[key + aggregation_field.name] = aggregation_field
                yield aggregation_field
            yield field

    @override
    def factory(
        self,
        model: type[T],
        dto_config: DTOConfig,
        base: type[Any] | None = None,
        name: str | None = None,
        parent_field_def: DTOFieldDefinition[ModelT, ModelFieldT] | None = None,
        current_node: Node[Relation[Any, GraphQLDTOT], None] | None = None,
        raise_if_no_fields: bool = False,
        backend_kwargs: dict[str, Any] | None = None,
        *,
        aggregations: bool = True,
        **kwargs: Any,
    ) -> type[GraphQLDTOT]:
        return super().factory(
            model,
            dto_config,
            base,
            name,
            parent_field_def,
            current_node,
            raise_if_no_fields,
            backend_kwargs,
            aggregations=aggregations if dto_config.purpose is Purpose.READ else False,
            **kwargs,
        )


class RootAggregateTypeDTOFactory(TypeDTOFactory[ModelT, ModelFieldT, GraphQLDTOT]):
    def __init__(
        self,
        inspector: GraphQLInspectorProtocol[Any, ModelFieldT],
        backend: DTOBackend[GraphQLDTOT],
        handle_cycles: bool = True,
        type_map: dict[Any, Any] | None = None,
        type_factory: TypeDTOFactory[ModelT, ModelFieldT, GraphQLDTOT] | None = None,
        aggregation_factory: AggregateDTOFactory[ModelT, ModelFieldT, AggregateDTOT] | None = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(inspector, backend, handle_cycles, type_map, **kwargs)
        self._type_factory = type_factory or TypeDTOFactory(inspector, backend)
        self._aggregation_factory = aggregation_factory or AggregateDTOFactory(
            inspector, DataclassDTOBackend(AggregateDTO)
        )

    @override
    def dto_name(
        self, base_name: str, dto_config: DTOConfig, node: Node[Relation[Any, GraphQLDTOT], None] | None = None
    ) -> str:
        return f"{base_name}Root"

    @override
    def iter_field_definitions(
        self,
        name: str,
        model: type[T],
        dto_config: DTOConfig,
        base: type[DTOBase[ModelT]] | None,
        node: Node[Relation[ModelT, GraphQLDTOT], None],
        raise_if_no_fields: bool = False,
        aggregations: bool = False,
        field_map: dict[DTOKey, GraphQLFieldDefinition[Any, Any]] | None = None,
        **kwargs: Any,
    ) -> Generator[DTOFieldDefinition[Any, ModelFieldT], None, None]:
        if not node.is_root:
            yield from ()
        key = DTOKey.from_dto_node(node)
        field_map = field_map if field_map is not None else {}
        nodes_dto = self._type_factory.factory(model, dto_config=dto_config, aggregations=aggregations)
        nodes = GraphQLFieldDefinition(
            dto_config=dto_config,
            model=model,
            model_field_name=NODES_KEY,
            type_hint=list[nodes_dto],
            is_relation=False,
        )
        aggregations_field = GraphQLFieldDefinition(
            dto_config=dto_config,
            model=model,
            model_field_name=AGGREGATIONS_KEY,
            type_hint=self._aggregation_factory.factory(model, dto_config=dto_config),
            is_relation=False,
            is_aggregate=True,
        )
        field_map[key + nodes.name] = nodes
        field_map[key + aggregations_field.name] = aggregations_field
        yield from iter((nodes, aggregations_field))

    @override
    def factory(
        self,
        model: type[T],
        dto_config: DTOConfig,
        base: type[Any] | None = None,
        name: str | None = None,
        parent_field_def: DTOFieldDefinition[ModelT, ModelFieldT] | None = None,
        current_node: Node[Relation[Any, GraphQLDTOT], None] | None = None,
        raise_if_no_fields: bool = False,
        backend_kwargs: dict[str, Any] | None = None,
        *,
        aggregations: bool = True,
        **kwargs: Any,
    ) -> type[GraphQLDTOT]:
        dto = super().factory(
            model,
            dto_config,
            base,
            name,
            parent_field_def,
            current_node,
            raise_if_no_fields,
            backend_kwargs,
            aggregations=aggregations,
            **kwargs,
        )
        dto.__strawchemy_is_root_aggregation_type__ = True
        return dto


class AggregateDTOFactory(GraphQLDTOFactory[ModelT, ModelFieldT, AggregateDTOT]):
    def __init__(
        self,
        inspector: GraphQLInspectorProtocol[Any, ModelFieldT],
        backend: DTOBackend[AggregateDTOT],
        handle_cycles: bool = True,
        type_map: dict[Any, Any] | None = None,
        aggregation_builder: AggregationInspector[ModelT, ModelFieldT] | None = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(inspector, backend, handle_cycles, type_map, **kwargs)
        self._aggregation_builder = aggregation_builder or AggregationInspector(inspector)

    @override
    def type_description(self) -> str:
        return "Aggregation fields"

    @override
    def dto_name(
        self, base_name: str, dto_config: DTOConfig, node: Node[Relation[Any, AggregateDTOT], None] | None = None
    ) -> str:
        return f"{base_name}Aggregate"

    @override
    def _factory(
        self,
        name: str,
        model: type[T],
        dto_config: DTOConfig,
        node: Node[Relation[Any, AggregateDTOT], None],
        base: type[Any] | None = None,
        parent_field_def: DTOFieldDefinition[ModelT, ModelFieldT] | None = None,
        raise_if_no_fields: bool = False,
        backend_kwargs: dict[str, Any] | None = None,
        field_map: dict[DTOKey, GraphQLFieldDefinition[Any, Any]] | None = None,
        **kwargs: Any,
    ) -> type[AggregateDTOT]:
        field_map = field_map if field_map is not None else {}
        model_field = parent_field_def.model_field if parent_field_def else None
        as_partial_config = dataclasses.replace(dto_config, partial=True)
        field_definitions: list[FunctionFieldDefinition[T, ModelFieldT]] = [
            FunctionFieldDefinition[T, ModelFieldT](
                dto_config=dto_config,
                model=model,
                _model_field=model_field if model_field is not None else DTO_MISSING,
                model_field_name=aggregation.function,
                type_hint=aggregation.output_type,
                _function=aggregation,
                default=aggregation.default,
            )
            for aggregation in self._aggregation_builder.output_functions(model, as_partial_config)
        ]

        root_key = DTOKey.from_dto_node(node)
        field_map.update({root_key + field.model_field_name: field for field in field_definitions})
        return self.backend.build(name, model, field_definitions, **(backend_kwargs or {}))


class DistinctOnFieldsDTOFactory(EnumDTOFactory[ModelT, ModelFieldT]):
    @override
    def dto_name(
        self, base_name: str, dto_config: DTOConfig, node: Node[Relation[Any, EnumDTO], None] | None = None
    ) -> str:
        return f"{base_name}DistinctOnFields"
