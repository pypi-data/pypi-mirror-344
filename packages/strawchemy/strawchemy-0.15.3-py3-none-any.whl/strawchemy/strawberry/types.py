from __future__ import annotations

from enum import Enum
from typing import TYPE_CHECKING, Any, ClassVar, TypeVar

import strawberry
from strawberry import UNSET
from strawberry.types import get_object_definition
from strawchemy.dto.base import MappedDTO
from strawchemy.graphql.mutation import (
    RequiredToManyUpdateInputMixin,
    RequiredToOneInputMixin,
    ToManyCreateInputMixin,
    ToManyUpdateInputMixin,
    ToOneInputMixin,
)
from strawchemy.utils import snake_to_lower_camel_case

if TYPE_CHECKING:
    from pydantic import ValidationError
    from pydantic_core import ErrorDetails


T = TypeVar("T", bound=MappedDTO[Any])
RelationInputT = TypeVar("RelationInputT", bound=MappedDTO[Any])
_TO_ONE_DESCRIPTION = "Add a new or existing object"
_TO_MANY_DESCRIPTION = "Add new or existing objects"
_TO_MANY_UPDATE_DESCRIPTION = "Add new objects or update existing ones"


def error_type_names() -> set[str]:
    return {get_object_definition(type_, strict=True).name for type_ in ErrorType.__error_types__}


class ErrorId(Enum):
    ERROR = "ERROR"
    VALIDATION_ERROR = "VALIDATION_ERROR"
    LOCALIZED_VALIDATION_ERROR = "LOCALIZED_VALIDATION_ERROR"


@strawberry.input(description=_TO_ONE_DESCRIPTION)
class ToOneInput(ToOneInputMixin[T, RelationInputT]):
    set: T | None = UNSET
    create: RelationInputT | None = UNSET


@strawberry.input(description=_TO_ONE_DESCRIPTION)
class RequiredToOneInput(RequiredToOneInputMixin[T, RelationInputT]):
    set: T | None = UNSET
    create: RelationInputT | None = UNSET


@strawberry.input(description=_TO_MANY_DESCRIPTION)
class ToManyCreateInput(ToManyCreateInputMixin[T, RelationInputT]):
    set: list[T] | None = UNSET
    add: list[T] | None = UNSET
    create: list[RelationInputT] | None = UNSET


@strawberry.input(description=_TO_MANY_UPDATE_DESCRIPTION)
class ToManyUpdateInput(ToManyUpdateInputMixin[T, RelationInputT]):
    set: list[T] | None = UNSET
    add: list[T] | None = UNSET
    remove: list[T] | None = UNSET
    create: list[RelationInputT] | None = UNSET


@strawberry.input(description=_TO_MANY_UPDATE_DESCRIPTION)
class RequiredToManyUpdateInput(RequiredToManyUpdateInputMixin[T, RelationInputT]):
    set: list[T] | None = UNSET
    add: list[T] | None = UNSET
    create: list[RelationInputT] | None = UNSET


@strawberry.interface(description="Base interface for expected errors", name="ErrorType")
class ErrorType:
    """Base class for GraphQL errors."""

    __error_types__: ClassVar[set[type[Any]]] = set()

    id: str = ErrorId.ERROR.value

    def __init_subclass__(cls) -> None:
        if not cls.__error_types__:
            cls.__error_types__.add(ErrorType)
        cls.__error_types__.add(cls)


@strawberry.type(description="Indicate validation error type and location.", name="LocalizedErrorType")
class LocalizedErrorType(ErrorType):
    """Match inner shape of pydantic ValidationError."""

    id = ErrorId.LOCALIZED_VALIDATION_ERROR.value
    loc: list[str] = strawberry.field(default_factory=list)
    message: str
    type: str


@strawberry.type(description="Input is malformed or invalid.", name="ValidationErrorType")
class ValidationErrorType(ErrorType):
    """Input is malformed or invalid."""

    id = ErrorId.VALIDATION_ERROR.value
    errors: list[LocalizedErrorType]

    @classmethod
    def _to_localized_error(cls, errors: ErrorDetails, to_camel: bool) -> LocalizedErrorType:
        return LocalizedErrorType(
            loc=[snake_to_lower_camel_case(str(loc)) if to_camel else str(loc) for loc in errors["loc"]],
            message=errors["msg"],
            type=errors["type"],
        )

    @classmethod
    def from_pydantic(cls, exc: ValidationError, to_camel: bool = True) -> ValidationErrorType:
        return ValidationErrorType(errors=[cls._to_localized_error(err, to_camel) for err in exc.errors()])
