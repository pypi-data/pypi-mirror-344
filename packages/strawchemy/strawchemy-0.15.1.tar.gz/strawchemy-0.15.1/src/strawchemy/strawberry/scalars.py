from __future__ import annotations

from datetime import timedelta
from functools import partial
from typing import NewType

from pydantic import TypeAdapter

from strawberry import scalar

__all__ = ("Interval",)

_IntervalType = TypeAdapter(timedelta)

Interval = scalar(
    NewType("Interval", timedelta),
    description=(
        "The `Interval` scalar type represents a duration of time as specified by "
        "[ISO 8601](https://en.wikipedia.org/wiki/ISO_8601#Durations)."
    ),
    parse_value=_IntervalType.validate_python,
    serialize=partial(_IntervalType.dump_python, mode="json"),
    specified_by_url="https://en.wikipedia.org/wiki/ISO_8601#Durations",
)
