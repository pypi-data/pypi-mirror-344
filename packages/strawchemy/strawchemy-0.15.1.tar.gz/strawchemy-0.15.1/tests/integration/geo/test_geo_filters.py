from __future__ import annotations

import pytest

pytest.importorskip("geoalchemy2", reason="geoalchemy2 is not installed")


from typing import TYPE_CHECKING, Any

from strawchemy import StrawchemyAsyncRepository, StrawchemySyncRepository

import strawberry
from sqlalchemy import Executable, Insert, MetaData, insert, text
from tests.integration.fixtures import QueryTracker
from tests.integration.utils import to_graphql_representation
from tests.typing import AnyQueryExecutor
from tests.utils import maybe_async

from .models import GeoModel, geo_metadata
from .types import GeoFieldsFilter, GeoFieldsType, strawchemy

if TYPE_CHECKING:
    from pytest_databases.docker.postgres import PostgresService

    from syrupy.assertion import SnapshotAssertion
    from tests.integration.typing import RawRecordData


pytestmark = [pytest.mark.integration, pytest.mark.geo]


@pytest.fixture
def metadata() -> MetaData:
    return geo_metadata


@pytest.fixture
def database_service(postgis_service: PostgresService) -> PostgresService:
    return postgis_service


@pytest.fixture
def before_create_all_statements() -> list[Executable]:
    return [text("CREATE EXTENSION IF NOT EXISTS postgis")]


@strawberry.type
class AsyncQuery:
    geo_field: list[GeoFieldsType] = strawchemy.field(
        filter_input=GeoFieldsFilter, repository_type=StrawchemyAsyncRepository
    )


@strawberry.type
class SyncQuery:
    geo_field: list[GeoFieldsType] = strawchemy.field(
        filter_input=GeoFieldsFilter, repository_type=StrawchemySyncRepository
    )


@pytest.fixture
def sync_query() -> type[SyncQuery]:
    return SyncQuery


@pytest.fixture
def async_query() -> type[AsyncQuery]:
    return AsyncQuery


@pytest.fixture
def seed_insert_statements(raw_geo: RawRecordData) -> list[Insert]:
    return [insert(GeoModel).values(raw_geo)]


@pytest.mark.snapshot
async def test_no_filtering(
    any_query: AnyQueryExecutor, raw_geo: RawRecordData, query_tracker: QueryTracker, sql_snapshot: SnapshotAssertion
) -> None:
    """Test that querying without filters returns all records."""
    result = await maybe_async(any_query("{ geoField { id } }"))
    assert not result.errors
    assert result.data
    assert len(result.data["geoField"]) == len(raw_geo)
    # Verify SQL query
    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.parametrize(
    ("field_name", "geometry", "expected_ids"),
    [
        pytest.param("polygon", {"type": "Point", "coordinates": [0.5, 0.5]}, [0, 3], id="point-within-polygon"),
        pytest.param("multiPolygon", {"type": "Point", "coordinates": [3, 3]}, [3], id="point-within-multipolygon"),
        pytest.param("geometry", {"type": "Point", "coordinates": [150, 150]}, [3], id="point-within-geometry-polygon"),
    ],
)
@pytest.mark.snapshot
async def test_contains_geometry(
    field_name: str,
    geometry: dict[str, Any],
    expected_ids: list[int],
    any_query: AnyQueryExecutor,
    raw_geo: RawRecordData,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
) -> None:
    """Test the contains_geometry filter.

    This tests that a geometry contains the provided geometry.
    For example, a polygon contains a point if the point is inside the polygon.
    """
    query = f"""
            {{
                geoField(filter: {{ {field_name}: {{ containsGeometry: {to_graphql_representation(geometry, "input")} }} }}) {{
                    id
                    {field_name}
                }}
            }}
    """
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data
    assert len(result.data["geoField"]) == len(expected_ids)
    for i, expected_id in enumerate(expected_ids):
        assert result.data["geoField"][i]["id"] == raw_geo[expected_id]["id"]
    # Verify SQL query
    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.parametrize(
    ("field_name", "geometry", "expected_ids"),
    [
        pytest.param(
            "point",
            {"type": "Polygon", "coordinates": [[[0, 0], [0, 2], [2, 2], [2, 0], [0, 0]]]},
            [0],
            id="point-within-polygon",
        ),
        pytest.param(
            "lineString",
            {"type": "Polygon", "coordinates": [[[0, 0], [0, 3], [3, 3], [3, 0], [0, 0]]]},
            [0],
            id="linestring-within-polygon",
        ),
        pytest.param(
            "multiPoint",
            {
                "coordinates": [
                    [
                        [3.8530909369223423, 3.5077205177229587],
                        [-2.126498556883888, 3.5077205177229587],
                        [-2.126498556883888, -0.8070832671228061],
                        [3.8530909369223423, -0.8070832671228061],
                        [3.8530909369223423, 3.5077205177229587],
                    ]
                ],
                "type": "Polygon",
            },
            [0],
            id="multipoint-within-polygon",
        ),
    ],
)
@pytest.mark.snapshot
async def test_within_geometry(
    field_name: str,
    geometry: dict[str, Any],
    expected_ids: list[int],
    any_query: AnyQueryExecutor,
    raw_geo: RawRecordData,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
) -> None:
    """Test the within_geometry filter.

    This tests that a geometry is within the provided geometry.
    For example, a point is within a polygon if the point is inside the polygon.
    """
    query = f"""
            {{
                geoField(filter: {{ {field_name}: {{ withinGeometry: {to_graphql_representation(geometry, "input")} }} }}) {{
                    id
                    {field_name}
                }}
            }}
    """
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data
    assert len(result.data["geoField"]) == len(expected_ids)
    for i, expected_id in enumerate(expected_ids):
        assert result.data["geoField"][i]["id"] == raw_geo[expected_id]["id"]
    # Verify SQL query
    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.snapshot
async def test_is_null(
    any_query: AnyQueryExecutor, raw_geo: RawRecordData, query_tracker: QueryTracker, sql_snapshot: SnapshotAssertion
) -> None:
    """Test the isNull filter for geometry fields."""
    query = """
            {
                geoField(filter: { point: { isNull: true } }) {
                    id
                    point
                }
            }
    """
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data
    assert len(result.data["geoField"]) == 1
    assert result.data["geoField"][0]["id"] == raw_geo[1]["id"]
    assert result.data["geoField"][0]["point"] is None
    # Verify SQL query
    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot
