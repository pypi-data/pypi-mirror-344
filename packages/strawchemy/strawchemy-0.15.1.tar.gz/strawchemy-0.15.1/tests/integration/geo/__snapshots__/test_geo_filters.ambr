# serializer version: 1
# name: test_contains_geometry[session-tracked-point-within-geometry-polygon-async-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.geometry) AS geometry,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.geometry, ST_GeomFromGeoJSON($1::VARCHAR))
  '''
# ---
# name: test_contains_geometry[session-tracked-point-within-geometry-polygon-async-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.geometry) AS geometry,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.geometry, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
# name: test_contains_geometry[session-tracked-point-within-geometry-polygon-sync-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.geometry) AS geometry,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.geometry, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
# name: test_contains_geometry[session-tracked-point-within-multipolygon-async-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.multi_polygon) AS multi_polygon,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.multi_polygon, ST_GeomFromGeoJSON($1::VARCHAR))
  '''
# ---
# name: test_contains_geometry[session-tracked-point-within-multipolygon-async-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.multi_polygon) AS multi_polygon,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.multi_polygon, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
# name: test_contains_geometry[session-tracked-point-within-multipolygon-sync-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.multi_polygon) AS multi_polygon,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.multi_polygon, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
# name: test_contains_geometry[session-tracked-point-within-polygon-async-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.polygon) AS
  POLYGON,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.polygon, ST_GeomFromGeoJSON($1::VARCHAR))
  '''
# ---
# name: test_contains_geometry[session-tracked-point-within-polygon-async-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.polygon) AS
  POLYGON,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.polygon, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
# name: test_contains_geometry[session-tracked-point-within-polygon-sync-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.polygon) AS
  POLYGON,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Contains(geos_fields.polygon, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
# name: test_is_null[session-tracked-async-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE geos_fields.point IS NULL
  '''
# ---
# name: test_is_null[session-tracked-async-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE geos_fields.point IS NULL
  '''
# ---
# name: test_is_null[session-tracked-sync-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE geos_fields.point IS NULL
  '''
# ---
# name: test_no_filtering[session-tracked-async-asyncpg_engine]
  '''
  SELECT geos_fields.id
    FROM geos_fields AS geos_fields
  '''
# ---
# name: test_no_filtering[session-tracked-async-psycopg_async_engine]
  '''
  SELECT geos_fields.id
    FROM geos_fields AS geos_fields
  '''
# ---
# name: test_no_filtering[session-tracked-sync-psycopg_engine]
  '''
  SELECT geos_fields.id
    FROM geos_fields AS geos_fields
  '''
# ---
# name: test_within_geometry[session-tracked-linestring-within-polygon-async-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.line_string) AS line_string,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.line_string, ST_GeomFromGeoJSON($1::VARCHAR))
  '''
# ---
# name: test_within_geometry[session-tracked-linestring-within-polygon-async-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.line_string) AS line_string,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.line_string, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
# name: test_within_geometry[session-tracked-linestring-within-polygon-sync-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.line_string) AS line_string,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.line_string, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
# name: test_within_geometry[session-tracked-multipoint-within-polygon-async-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.multi_point) AS multi_point,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.multi_point, ST_GeomFromGeoJSON($1::VARCHAR))
  '''
# ---
# name: test_within_geometry[session-tracked-multipoint-within-polygon-async-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.multi_point) AS multi_point,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.multi_point, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
# name: test_within_geometry[session-tracked-multipoint-within-polygon-sync-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.multi_point) AS multi_point,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.multi_point, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
# name: test_within_geometry[session-tracked-point-within-polygon-async-asyncpg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.point, ST_GeomFromGeoJSON($1::VARCHAR))
  '''
# ---
# name: test_within_geometry[session-tracked-point-within-polygon-async-psycopg_async_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.point, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
# name: test_within_geometry[session-tracked-point-within-polygon-sync-psycopg_engine]
  '''
  SELECT ST_AsEWKB(geos_fields.point) AS POINT,
         geos_fields.id
    FROM geos_fields AS geos_fields
   WHERE ST_Within(geos_fields.point, ST_GeomFromGeoJSON(%(ST_GeomFromGeoJSON_1)s::VARCHAR))
  '''
# ---
